### PR TITLE
mssql/msdsn: break out DSN parser from the rest of the package

### PR DESCRIPTION
**Pull request by kardianos**
This allows creating a Connector without parsing the DSN.

### CODE REVIEW

**Comment by kardianos** [comment]
@wrosenuance @denisenkom @coquagli

Here's what I'm thinking to expose the DSN parsing.

@wrosenuance by returning the raw parameters from parse as well, you can add extra parameters when using federated auth.

I also pulled out the certificate as an entity, so the driver doesn't need to hit the disk, but it could be in a configuration as well.

File: fedauth.go [hunk 1]
```diff
@@ -1,82 +1,82 @@
...
-	if tokenProvider == nil {
-		return nil, errors.New("mssql: tokenProvider cannot be nil")
-	}
-
-	conn, err := NewConnector(dsn)
-	if err != nil {
-		return nil, err
-	}
-
-	conn.params.fedAuthLibrary = fedAuthLibrarySecurityToken
```
Comment (wrosenuance) [inline fedauth.go:52]: If it's okay to add the ADAL workflow as well as the FedAuthLibrary to the new Config object, we can ditch these two connector creators in favor of NewConnectorConfig.

File: msdsn/conn_str.go [hunk 2]
```diff
@@ -11,55 +14,97 @@ import (
...
+	ReadOnlyIntent bool
+
+	LogFlags Log
+
+	ServerSPN   string
+	Workstation string
+	AppName     string
+
+	FedAuthLibrary int
+
```
Comment (wrosenuance) [inline msdsn/conn_str.go:65]: Could you add back FedAuthADALWorkflow (or just named ADALWorkflow) as a byte here? Would avoid the need to then have the same fields in the connector and the special newSecurityTokenConnector and newActiveDirectoryTokenConnector.

File: mssql.go [hunk 3]
```diff
@@ -92,15 +97,27 @@ func NewConnector(dsn string) (*Connector, error) {
...
 // at any time.
 //
 // In the future, settings that cannot be passed through a string DSN
 // may be set directly on the connector.
 type Connector struct {
-	params connectParams
+	params msdsn.Config
 	driver *Driver
 
+	fedAuthLibrary      int
```
Comment (wrosenuance) [inline mssql.go:118]: These could be removed if ADAL workflow is added to msdsn.Config

**Review by wrosenuance — COMMENTED** [review]
This all looks good to me - I can easily adapt the rest of the AD auth PR to use this.

I made several comments in places where it came up - essentially the only thing I would change is to expose the ADAL workflow byte alongside the FedAuthLibrary in the new Config struct, then we don't need to set them on the Connector and that cuts out a bunch of code.

**Comment by kardianos** [comment]
I'll look at the `fedAuthLbrary` zero value issue. Goal here is to continue to not expose any new API (other then existing and new msdsn), but get it situated for the next commit with fed auth. So I'm okay with a few params in the connector that are sub-optimally placed for now.

If we do expose something on the Config, I would want it to be a package controlled "enum" const, like what I did with Encryption, that would then get translated over to the correct values. I'm okay with adding that if it is package, not protocol, defined.

**Comment by wrosenuance** [comment]
> Goal here is to continue to not expose any new API (other then existing and new msdsn), but get it situated for the next commit with fed auth. So I'm okay with a few params in the connector that are sub-optimally placed for now.

After a little more reflection I think it makes more sense to have them in the Connector anyway, since...

> I'm okay with adding that if it is package, not protocol, defined.

... and these are protocol defined constants that don't really make sense as arbitrarily specified connection parameters.

> I'll look at the `fedAuthLbrary` zero value issue.

Perhaps the most straightforward thing is to have a bool flag in the Connector struct that has to be set to true for the fed auth to activate? That way it would default to off unless one of the specific Connector creation functions was called. Then those places where the flag is set to true can also make sure to set fedAuthLibrary correctly.

**Comment by wrosenuance** [comment]
I was curious as to why the tests didn't fail with the fedAuthLibrary defaulting to zero, so I checked out your branch to test. I think it's because that's the code for the Live ID token scheme and it's not supported, so attempting to use it falls through to other mechanisms.

I added a commit on my branch where I use an explicit flag to indicate that the fedAuthLibrary value has been initialized - https://github.com/denisenkom/go-mssqldb/compare/msdsn...wrosenuance:msdsn

**Comment by wrosenuance** [comment]
Hi @kardianos, is there something I can do to help with this part of the PR? I'd like to finish the rest of the Azure authentication PR, but I think this needs to be finalized first.

**Comment by kardianos** [comment]
@wrosenuance Let me try to finish this and address your comments. I'm sorry for the monumental delay.

**Pull request merged**